c10/core/Casting.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <c10/core/ScalarType.h>
+#include <c10/macros/Macros.h>
+#include <c10/util/Exception.h>
+#include <c10/util/Optional.h>
+
+#include <iostream>
+
+namespace c10 {
+
+/**
+ * Casting option for converting between types.
+ */
+enum class Casting : int8_t { No, Promote, Unsafe, NumOptions };
+
+static inline bool canCast(ScalarType from, ScalarType to, Casting casting) {
+  switch (casting) {
+    case Casting::No:
+      return (to == from);
+    case Casting::Promote:
+      return (to == promoteTypes(from, to));
```

Comment:
`np.promote_types` and thus `np.result_type` is using `safe` promotion. The operators then use `casting` to determine whether a type coercion is allowed from result type to output type. In that sense, 'safe' is NumPy's no coercion option, while 'same_kind' is the default coercion rule.

For any situation where there is no need for coercion (no result/output Tensor provided), the behavior of `promote_types` is used, which is 'safe` in NumPy.

### Regarding promote

`torch.promote_types` uses a different behavior than `np.promote_types` and it allows by itself coercion into smaller precision (but not smaller range). For example, operating on int64 and float32 yields float32, which is loss of information from the first input. I'm reluctant to call this 'safe' for this sole reason. I also don't think we should have both 'safe' and 'promote' at the same time.

This is a question of whether casting enum should include the default behavior of `promote_types`, which will be the casting behavior for all cases where there is no target output tensor. I'm leaning towards including because this is *a* casting behavior used in he library and it's better if casting includes it for the least surprise.

### Regarding same_kind

Since `same_kind` is sharing some behavior with `promote`, it makes the API a bit unorthogonal.

With `same_kind`,
- if my aim was to promote integers to floats without concern for precision (int64 -> float32), `torch.promote_types` and thus `promote` already handles that.
- If my aim was to specifically allow precision loss (float64 -> float32), then 'unsafe' covers that. 

### Options

1. **[no, same_kind, unsafe]** (your suggestion). This leaves the no coercion casting to be something very similar to `same_kind` ('promote' here) but leave it out of the enum. The clunkiness isn't resolved by just removing the option because it's going be the type casting behavior for quite a significant portion of calls, at the same time introducing a very similar new casting rule (same_kind).
2. **[no, promote, same_kind, unsafe]**: `promote` and `same_kind` are very similar to each other. Since `promote` already allows some coercion, I don't there is any point in introducing `same_kind` other than be compatible with NumPy in default coercion behavior.
3. **[no, safe, same_kind, unsafe]** (the clean way out). This is possible with changing default promotion through `promote_types` be `safe` (same as NumPy). As we discussed before, that would lean towards more casts to float64 when integer tensors are involved.
4. **[no, promote, unsafe]** (current version here). This is the one that deviates from NumPy the most but it's the most coherent if we want to keep `torch.promote_types` deviating from `np.promote_types`.

I'm leaning a lot towards 3 but you have more context on the decision around `promote_types` so let me know what you think.

Reply:
I'm not really sure what this means:
"`np.promote_types` and thus `np.result_type` is using safe promotion"

As best I can tell, this means `np.promote_types(a,b)` always satisfies:
`np.can_cast(a, b,' safe') || np.can_cast(b,a, 'safe')`

but this doesn't follow for `torch.promote_types(float32, in64)` because:
`torch.promote_types(float32, in64) -> float32`
but:
```
torch.can_cast(float32, int64) -> False
torch.can_cast(float64, int32) -> False
```

is that right?

Reply:
> As best I can tell, this means np.promote_types(a,b) always satisfies:
> np.can_cast(a, b,' safe') || np.can_cast(b,a, 'safe')

That's not quite right. It satisfies:
```
c = np.promote_types(a,b)
np.can_cast(a, c,'safe') && np.can_cast(b, c, 'safe')
```
The promoted is always a safe cast from either input.

Float cannot be cast into integer either in NumPy or in PyTorch. Because that would losing range, which is unsafe for both 'safe' and 'promote'.

The difference is how willing either is to lose precision.

Numpy 'safe' tries to preserve each and every value when allowing a safe cast.

```
np.promote_types(float32, int32) -> float64 (use a float whose mantissa can fit an int32)
np.can_cast(int32, float32) -> False (loses precision, int32 cannot fit in mantissa of float32)
np.can_cast(float32, int64) -> False (loses range)
```

Current PR (default is 'promote')
```
torch.promote_types(float16, int32) -> float16 (any integer can be promoted to any float in PyTorch)
torch.can_cast(int32, float16) -> True (ignoring precision again)
torch.can_cast(float32, int64) -> False (loses range, PyTorch doesn't allow that for any circumstance)
```

NumPy's default promote_types (and thus 'safe' can_cast) cares about precision but PyTorch doesn't.

There is a quirk in NumPy where it foregoes precision once you reach float64, even though it has float128.
```
torch.promote_types(float64, int64) -> float64
torch.can_cast(int64, float64) -> True (losing precision, some int64 values will change)
```

I couldn't find the reasoning behind this but it looks like a backwards compatibility decision. PyTorch's quirk is the same in nature, except it occurs right at the smaller float which is float16.

- Given that NumPy has a similar behavior between float32 and float64, I think it's fine to keep promote_types in PyTorch as is and still call it 'safe'
- An alternative would be to try to preserve precision until float32 (which is the default type) but forego precision beyond that (NumPy does this at float64). In that case, nothing would get silently promoted to float64 but you'd have to be explicit about keeping things in float16 (which is already the case for the tensor creation ops).

Reply:
Ok this makes sense, thanks for the writeup.  I'm still not sold that we actually need to do this now vs say just implement 'same_kind' only and not expose it.